### PR TITLE
Addressed a PocketBook SDK v6 breaking change related to HID events

### PR DESCRIPTION
**Pull request by tommiv**
Discussion is in https://github.com/koreader/koreader/issues/12703

- - -
This change is [Reviewable](https://reviewable.io/reviews/koreader/koreader-base/1970)

### CODE REVIEW

**Comment by tommiv** [comment]
@Frenzie yes. I've just dropped a more detailed comment here https://github.com/koreader/koreader/issues/12703#issuecomment-2454538535

**Comment by benoit-pierre** [comment]
It would be better to update `ffi-cdecl/inkview_decl.c` and re-generate `ffi-cdecl/inkview_decl.c` with ffi-cdecl.

Do you have a build environment setup?

**Comment by benoit-pierre** [comment]
Can some pocketbooks be running on old firmware with the old definitions?

**Comment by tommiv** [comment]
> Do you have a build environment setup?
Unfortunately I don’t.

> Can some pocketbooks be running on old firmware with the old definitions?
Absolutely. Though it looks like SDK v6 is out and about for 2 years now

**Comment by tommiv** [comment]
It probably may be done in a more safe way if I react on _EXT events only when the device runs v6.x firmware, but unfortunately I have no idea how to check the version in runtime

**Comment by benoit-pierre** [comment]
There's `getSoftwareVersion()`, there should be a trace in the log at startup: `SoftwareVersion: …`, can you check yours?

**Comment by tommiv** [comment]
Indeed, the log entry looks like this:
```
11/04/24-13:45:08 INFO  SoftwareVersion:  U743g.6.8.4143
```

**Comment by benoit-pierre** [comment]
Have you actually checked that `par1 / par2` values don't overlap, between ext and non-ext events?

**Comment by NiLuJe** [comment]
> Have you actually checked that `par1 / par2` values don't overlap, between ext and non-ext events?

I suspect that they theoretically can, but IV has historically been using custom keycodes for its basic needs, AFAICT.

(c.f., https://github.com/pocketbook/SDK_6.3.0/blob/dd8de463733a2bc0d53d1214b64734586fac427e/SDK-B288/usr/arm-obreey-linux-gnueabi/sysroot/usr/local/include/inkview.h#L475-L529)

I'm going to assume that if it currently works, eeeh, should be fine.

Since we don't actually use the TOUCHUP/TOUCHDOWN/TOUCHMOVE stuff, the breaking API change shouldn't really concern us.

**Comment by NiLuJe** [comment]
> And since we don't actually use the TOUCHUP/TOUCHDOWN/TOUCHMOVE stuff, the breaking API change shouldn't really concern us either.

Unless these events are actually common on older FW, in which case, yeah, that's going to be a problem on FW < 6 devices ;D.

**Comment by tommiv** [comment]
I actually do hope they have the same values (ie linux virtual keycodes) can’t prove it though, there is no hardware keyboard on-device, only 4 functional buttons. However, for on-board buttons the event is still without _EXT postfix, while HID keyboards always use _EXT events. I wonder if this ext actually stands for “external” and not “extended” as I thought initially

**Comment by NiLuJe** [comment]
> I suspect that they theoretically can, but IV has historically been using custom keycodes for its basic needs, AFAICT.

I'm going by the assumption that the _EXT events are reserved for external input devices (e.g., OTG, BT), as opposed to the device's actual buttons for the non-_EXT variants.

**Comment by tommiv** [comment]
> Unless these events are actually common on older FW, in which case, yeah, that's going to be a problem on FW < 6 devices ;D.

this can be mitigated nevertheless, if new event codes are only consumed if fw major version is >= 6. Allow me some time, I’ll add the check

**Comment by benoit-pierre** [comment]
> > Have you actually checked that `par1 / par2` values don't overlap, between ext and non-ext events?
> 
> I suspect that they theoretically can, but IV has historically been using custom keycodes for its basic needs, AFAICT.
> 
> (c.f., https://github.com/pocketbook/SDK_6.3.0/blob/dd8de463733a2bc0d53d1214b64734586fac427e/SDK-B288/usr/arm-obreey-linux-gnueabi/sysroot/usr/local/include/inkview.h#L475-L529)
> 
> I'm going to assume that if it currently works, eeeh, should be fine.

Because in traces of https://github.com/koreader/koreader/issues/12703#issue-2631378050, there's:
```
11/03/24-18:41:09 DEBUG PB event => 26 (EVT_KEYRELEASE), par1: 24, par2: 0 
```
and
```
11/03/24-18:41:49 DEBUG PB event => 40 (EVT_TOUCHUP), par1: 23, par2: 0 
```
That's dangerously close…

**Comment by NiLuJe** [comment]
Oh, right, we got away with it because the old SDK doesn't use custom constants, but just undefs & redefs the Linux KEY_ ones...

But you're right, we need the new mapping with custom names, make non-_EXT only handle the IV_ constants, and EXT handle the standard Linux ones.

**Comment by NiLuJe** [comment]
i.e., compare https://github.com/blchinezu/pocketbook-sdk/blob/64e9fa210dd90569f8eabea92da4d40f3d072fc1/FRSCSDK/arm-none-linux-gnueabi/sysroot/usr/include/inkview.h#L283-L347 to https://github.com/pocketbook/SDK_6.3.0/blob/dd8de463733a2bc0d53d1214b64734586fac427e/SDK-B288/usr/arm-obreey-linux-gnueabi/sysroot/usr/local/include/inkview.h#L475-L529

(The good news is the actual values don't appear to have changed when they switched to a custom `IV_` namespace for the same constants in FW 5.x).

**Comment by NiLuJe** [comment]
Following these old, old headers in base translates to: https://github.com/koreader/koreader-base/blob/ddcec0afa634e590c584e1fb76cc7babeb63a284/ffi/inkview_h.lua#L815-L848

(And I'm guessing we avoided a conflict with our https://github.com/koreader/koreader-base/blob/master/ffi/linux_input_h.lua by blind luck ;p)

which meant front could happily shove this mess under the carpet in front: https://github.com/koreader/koreader/blob/1e55dda4c765be3316d7ca87be3052943cb1be86/frontend/device/pocketbook/device.lua#L191-L200

**Comment by NiLuJe** [comment]
> And then update the `event_map` in front to point to the `IV_` ones where relevant...

Or, perhaps saner for future readers, make the `non-_EXT` handler translate the `IV_`constants to their standard Linux counterparts, and update the event_map accordingly.

**Comment by tommiv** [comment]
Well guys, I'm completely lost in your conversation, looks like I bit more than I can swallow. I've added v6+ check nevertheless, but it seems you have better ideas of how to incorporate v6 changes into the codebase. Anyway, let me know if I can help.

**Comment by benoit-pierre** [comment]
@NiLuJe: that part seems really ugly:
```lua
        if band(poll_fds[0].revents, C.POLLIN) ~= 0 then
            updateTimestamp() -- single 'ts' copy for genEmuEvent inside the loop
            while rt.mq_receive(poll_fds[0].fd, ffi.cast("char*", hwmsg), hwmsg_len, nil) > 0 do
                local m = hwmsg[0]
                -- If there's no raw keymapping, emit this one instead
                if (m.type == C.EVT_KEYDOWN or m.type == C.EVT_KEYUP or m.type == C.EVT_KEYREPEAT) and not raw_keymap then
                    genEmuEvent(m.type, m.common.par1, m.common.par2)
                end
            end
        end
```

Since `EV_MAX` in `base/ffi/linux_input_h.lua` is 31, and those `EVT_KEYXXX` values start at 25.

Anyway, @tommiv, does this work:  [inkview_sdk_v6_update.zip](https://github.com/user-attachments/files/17620625/inkview_sdk_v6_update.zip)?

**Comment by tommiv** [comment]
@benoit-pierre let me revert my changes for a clear experiment, I'll be back soon

**Comment by tommiv** [comment]
Ah, there's no need to revert anything, files in the archive replacing ones I edited. Unfortunately, it crashes after the replacement:
```
 [*] Current time: 11/04/24-18:05:18
 [*] Version: v2024.07

ffi.load: libs/libutf8proc.so.3
ffi.load: blitbuffer
11/04/24-18:05:18 DEBUG fts: Preferred MONOTONIC clock source is CLOCK_MONOTONIC 
11/04/24-18:05:18 DEBUG fts: Preferred REALTIME clock source is CLOCK_REALTIME 
11/04/24-18:05:18 DEBUG fts: BOOTTIME clock source is supported 
ffi.load: inkview
./luajit: ./ffi/inkview_h.lua:141: size of C type is unknown or too large at line 69
stack traceback:
	[C]: in function 'cdef'
	./ffi/inkview_h.lua:141: in main chunk
	[C]: in function 'require'
	frontend/device/pocketbook/device.lua:14: in main chunk
	[C]: in function 'probeDevice'
	frontend/device.lua:61: in main chunk
	[C]: in function 'require'
	./reader.lua:151: in main chunk
	[C]: at 0x00013f61
!!!!
Uh oh, something went awry... (Crash n°5: 11/04/24 @ 18:05:18)
Too many consecutive crashes, aborting . . .
!!!! ! !!!!
```

**Comment by benoit-pierre** [comment]
OK, I screwed the FFI declarations, new version: 
[inkview_sdk_v6_update.zip](https://github.com/user-attachments/files/17620918/inkview_sdk_v6_update.zip)

**Comment by tommiv** [comment]
Still crashes, but this time the log is way longer than previous one, many things happen before the error. The error itself is also different:
```
11/04/24-18:19:11 DEBUG triggering refresh {
  mode = "full",
  region = "1404x1872+0+0"
} --[[table: 0xa1ed6e00]] 
11/04/24-18:19:11 DEBUG refresh: inkview full 0 0 1404 1872 nil 
./luajit: ./ffi/input_pocketbook.lua:429: missing declaration for symbol 'ProcessEventLoop'
stack traceback:
	[C]: in function '__index'
	./ffi/input_pocketbook.lua:429: in function 'waitForEvent'
	frontend/device/input.lua:1399: in function 'waitEvent'
	frontend/ui/uimanager.lua:1535: in function 'handleInput'
	frontend/ui/uimanager.lua:1590: in function 'run'
	./reader.lua:265: in main chunk
	[C]: at 0x00013f61
```

**Comment by benoit-pierre** [comment]
And I was missing the compat declarations: 
[inkview_sdk_v6_update.zip](https://github.com/user-attachments/files/17621045/inkview_sdk_v6_update.zip)

**Comment by NiLuJe** [comment]
> @NiLuJe: that part seems really ugly:

`waitForEventRaw` is barely used anywhere, as it both requires root, *and* a devcap flag, so, IIRC there's basically one device flagged as such: the one the author tested this on.

Add to that that this bypasses InkView entirely, and is actually polling an evdev device, so the InkView mapping is no longer relevant anyway ;).

**Comment by NiLuJe** [comment]
Eh, ignore the above, not on my machine, clearly I can't read ;p.

**Comment by benoit-pierre** [comment]
OK, did another pass on constants and added the missing declaration for `pocketbooksync.koplugin` [inkview_sdk_v6_update.zip](https://github.com/user-attachments/files/17621226/inkview_sdk_v6_update.zip).

**Comment by NiLuJe** [comment]
> Eh, ignore the above, not on my machine, clearly I can't read ;p.

Okay, maybe I wasn't *entirely* off. Sleep deprivation FTW, yay!

Anyway, yeah, not quite sure what this does, but it is barely used anywhere because of the root requirement. I assume this specific monitor fd only outputs a similar InkView-like mess like the standard codepath, but I really have no idea, since Ive never seen that codepath exercised anywhere.

**Comment by tommiv** [comment]
@benoit-pierre no errors anymore (including the one from the plugin). On-device buttons work, gamepad works too. Here are some logs just in case.

On-device "back" button (negative code is something new, but I assume it's your solution to separate non-EXT and EXT codes):
```
11/04/24-18:44:01 DEBUG PB event => 25 (EVT_KEYPRESS), par1: 24, par2: 0 
11/04/24-18:44:01 DEBUG key event => code: -24 (LPgBack), value: 1, time: 1730738668.124038 
...
11/04/24-18:44:01 DEBUG PB event => 26 (EVT_KEYRELEASE), par1: 24, par2: 0 
11/04/24-18:44:01 DEBUG key event => code: -24 (LPgBack), value: 0, time: 1730738668.284355 
```
Same behavior for on-device "forward" button, just a different code (-25)

BT gamepad "dpad-up" button (mapped in input_map.lua):
```
11/04/24-18:44:06 DEBUG PB event => 40 (EVT_KEYPRESS_EXT), par1: 46, par2: 0 
11/04/24-18:44:06 DEBUG key event => code: 46 (LPgFwd), value: 1, time: 1730738673.856902 
...
11/04/24-18:44:06 DEBUG PB event => 41 (EVT_KEYRELEASE_EXT), par1: 46, par2: 0 
11/04/24-18:44:06 DEBUG key event => code: 46 (LPgFwd), value: 0, time: 1730738673.958813 
```
Same for "dpad-down", just a different code (32) and mapped command (LPgBack).

tl;dr: works like a charm!

**Comment by tommiv** [comment]
@benoit-pierre I guess I'm closing this PR in favor of your solution, since it's much more architecturally correct, and continue to run the locally patched version on my Pocketbook until the new koreader release is out, right?

**Comment by Frenzie** [comment]
> until the new koreader release is out, right?

The one after, I'd say. ;-)

**Comment by tommiv** [comment]
Closed in favor of https://github.com/koreader/koreader-base/pull/1971
